- Marsya Nur Iman Binti Mohd Ishak (IS01081187)
- Nur Amira Hanina Binti Khairul Nazmi (IS0101120)

In [2]:
import pandas as pd
import re
import string 
import nltk
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer

In [3]:
df = pd.read_csv("cleanReviews.csv")
df.head()

Id  Score                                       clean_number  \
0   1      5  i have bought several of the vitality canned d...   
1   2      1  product arrived labeled as jumbo salted peanut...   
2   3      4  this is a confection that has been around a fe...   
3   4      2  if you are looking for the secret ingredient i...   
4   5      5  great taffy at a great price  there was a wide...   

                                          token_data  
0  ['i', 'have', 'bought', 'several', 'of', 'the'...  
1  ['product', 'arrived', 'labeled', 'as', 'jumbo...  
2  ['this', 'is', 'a', 'confection', 'that', 'has...  
3  ['if', 'you', 'are', 'looking', 'for', 'the', ...  
4  ['great', 'taffy', 'at', 'a', 'great', 'price'...

In [4]:
df.shape

(568454, 4)

In [5]:
# read data
df = pd.read_csv("cleanReviews.csv")

In [6]:
df.head(10)

Id  Score                                       clean_number  \
0   1      5  i have bought several of the vitality canned d...   
1   2      1  product arrived labeled as jumbo salted peanut...   
2   3      4  this is a confection that has been around a fe...   
3   4      2  if you are looking for the secret ingredient i...   
4   5      5  great taffy at a great price  there was a wide...   
5   6      4  i got a wild hair for taffy and ordered this f...   
6   7      5  this saltwater taffy had great flavors and was...   
7   8      5  this taffy is so good  it is very soft and che...   
8   9      5  right now im mostly just sprouting this so my ...   
9  10      5  this is a very healthy dog food good for their...   

                                          token_data  
0  ['i', 'have', 'bought', 'several', 'of', 'the'...  
1  ['product', 'arrived', 'labeled', 'as', 'jumbo...  
2  ['this', 'is', 'a', 'confection', 'that', 'has...  
3  ['if', 'you', 'are', 'looking', 'for', 'the', ...  
4  ['great', 'taffy', 'at', 'a', 'great', 'price'...  
5  ['i', 'got', 'a', 'wild', 'hair', 'for', 'taff...  
6  ['this', 'saltwater', 'taffy', 'had', 'great',...  
7  ['this', 'taffy', 'is', 'so', 'good', 'it', 'i...  
8  ['right', 'now', 'im', 'mostly', 'just', 'spro...  
9  ['this', 'is', 'a', 'very', 'healthy', 'dog', ...

In [7]:
df.shape

(568454, 4)

In [8]:
# select clean_number column
df_t = df['clean_number']
df_t.head()

0    i have bought several of the vitality canned d...
1    product arrived labeled as jumbo salted peanut...
2    this is a confection that has been around a fe...
3    if you are looking for the secret ingredient i...
4    great taffy at a great price  there was a wide...
Name: clean_number, dtype: object

In [9]:
# check no. of duplicates in clean_number column
df_t_duplicates = df_t.duplicated()
print(df_t_duplicates.sum())

174959


In [10]:
# remove duplicates in clean_number column
df_rdup = df_t.drop_duplicates(keep='first')
df_rdup.shape

(393495,)

# Removing HTML tags and unwanted characters

In [11]:
def get_cleaned_textdata(sentence):
    modified_sentence = re.sub(r'<.*?>',' ', sentence)
    modified_sentence = ''.join([i if i not in string.punctuation else ' ' for i in modified_sentence])
    modified_sentence = re.sub(r'\d+', ' ', modified_sentence)
    modified_sentence = re.sub(r'\s+', ' ', modified_sentence)
    modified_sentence = modified_sentence.lower()
    return modified_sentence

# Remove Duplicate Records

In [12]:
df_rdupclean = df_rdup.apply(get_cleaned_textdata)

In [13]:
# check any row
print(df_rdupclean[200])

even with small containers they dont fill them up these little tins are less than half filled and at the price charged it seems a ripoff is there some exotic ingredient as costly as gold contained in those tiny squares or how about the cereal ploy they were filled at the factory but settled in transportbr can manufacturers be honest in their dealings


In [14]:
print(df_rdupclean)

0         i have bought several of the vitality canned d...
1         product arrived labeled as jumbo salted peanut...
2         this is a confection that has been around a fe...
3         if you are looking for the secret ingredient i...
4         great taffy at a great price there was a wide ...
                                ...                        
568449    great for sesame chickenthis is a good if not ...
568450    im disappointed with the flavor the chocolate ...
568451    these stars are small so you can give of those...
568452    these are the best treats for training and rew...
568453    i am very satisfied product is as advertised i...
Name: clean_number, Length: 393495, dtype: object


In [15]:
# Select relevant columns for sentiment analysis
df = df[['Score', 'clean_number']]

# Feature Extraction (Bag Of Words AND TF-IDF)

In [16]:
from sklearn.feature_extraction.text import CountVectorizer, TfidfVectorizer

# Bag Of Words

In [17]:
# Initialize the CountVectorizer
vectorizer = CountVectorizer()

# Fit and transform the clean_number data
bow_features = vectorizer.fit_transform(df['clean_number'])

# Get the vocabulary (unique words)
vocabulary = vectorizer.get_feature_names_out()

# Print the shape of the BoW features and the vocabulary size
print("BoW feature shape:", bow_features.shape)
print("Vocabulary size:", len(vocabulary))

BoW feature shape: (568454, 224192)
Vocabulary size: 224192


In [18]:
from collections import Counter

# Given list of sentences
sentences = [
    "this is a very healthy dog food good for their digestion also good for small puppies my dog eats her required amount at every feeding",
    "great taffy at a great price  there was a wide assortment of yummy taffy  delivery was very quick  if your a taffy lover this is a deal",
    "this saltwater taffy had great flavors and was very soft and chewy  each candy was individually wrapped well  none of the candies were stuck together which did happen in the expensive version fralingers  would highly recommend this candy  i served it at a beachthemed party and everyone loved it",
    "this taffy is so good  it is very soft and chewy  the flavors are amazing  i would definitely recommend you buying it  very satisfying",
]

# Step 1: List all unique words
all_words = ' '.join(sentences).split()
unique_words = list(set(all_words))

# Step 2: Create word frequency table
word_freq = Counter(all_words)
sorted_word_freq = dict(sorted(word_freq.items(), key=lambda item: item[1], reverse=True))

# Step 3: Create document vectors
document_vectors = []
for sentence in sentences:
    sentence_words = sentence.split()
    document_vector = [1 if word in sentence_words else 0 for word in sorted_word_freq.keys()]
    document_vectors.append(document_vector)

# Print the results
print("Step 1: List of Unique Words:", unique_words)
print("\nStep 2: Word Frequency Table:")
print("Word\tFrequency")
for word, freq in sorted_word_freq.items():
    print(f"{word}\t{freq}")

print("\nStep 3: Document Vectors:")
for sentence, vector in zip(sentences, document_vectors):
    print(f"Sentence: {sentence}\nVector: {vector}\n")

Step 1: List of Unique Words: ['if', 'each', 'the', 'flavors', 'amazing', 'which', 'satisfying', 'and', 'also', 'there', 'soft', 'food', 'great', 'wide', 'so', 'would', 'beachthemed', 'happen', 'her', 'everyone', 'expensive', 'deal', 'good', 'puppies', 'you', 'were', 'dog', 'at', 'for', 'eats', 'definitely', 'candies', 'taffy', 'this', 'party', 'version', 'amount', 'a', 'fralingers', 'in', 'price', 'recommend', 'well', 'buying', 'are', 'every', 'required', 'of', 'small', 'digestion', 'stuck', 'individually', 'did', 'together', 'lover', 'had', 'it', 'loved', 'delivery', 'assortment', 'candy', 'yummy', 'chewy', 'saltwater', 'highly', 'their', 'was', 'your', 'wrapped', 'is', 'served', 'quick', 'feeding', 'my', 'i', 'very', 'healthy', 'none']

Step 2: Word Frequency Table:
Word	Frequency
a	6
this	5
very	5
taffy	5
is	4
was	4
and	4
it	4
good	3
at	3
great	3
the	3
dog	2
for	2
of	2
flavors	2
soft	2
chewy	2
candy	2
would	2
recommend	2
i	2
healthy	1
food	1
their	1
digestion	1
also	1
small	1
puppi

# Term Frequency-Inverse Document Frequency (TF-IDF)

In [19]:
# Initialize the TfidfVectorizer
tfidf_vectorizer = TfidfVectorizer()

# Fit and transform the preprocessed text data
tfidf_features = tfidf_vectorizer.fit_transform(df['clean_number'])

# Get the vocabulary (unique words)
tfidf_vocabulary = tfidf_vectorizer.get_feature_names_out()

# Print the shape of the TF-IDF features and the vocabulary size
print("TF-IDF feature shape:", tfidf_features.shape)
print("Vocabulary size:", len(tfidf_vocabulary))

TF-IDF feature shape: (568454, 224192)
Vocabulary size: 224192


In [20]:
import math
from collections import Counter

def compute_tf(document):
    word_count = Counter(document)
    tf = {word: count/len(document) for word, count in word_count.items()}
    return tf

def compute_idf(documents):
    N = len(documents)
    idf = {}
    all_words = set(word for doc in documents for word in doc)
    for word in all_words:
        count = sum(1 for doc in documents if word in doc)
        idf[word] = math.log(N/count)
    return idf

def compute_tfidf(document, idf):
    tfidf = {}
    tf = compute_tf(document)
    for word, tf_value in tf.items():
        tfidf[word] = tf_value * idf[word]
    return tfidf

# New data
data = [
    "this taffy is so good it is very soft and chewy the flavors are amazing i would definitely recommend you buying it very satisfying",
    "this is a very healthy dog food good for their digestion also good for small puppies my dog eats her required amount at every feeding",
    "this saltwater taffy had great flavors and was very soft and chewy  each candy was individually wrapped well  none of the candies were stuck together which did happen in the expensive version fralingers  would highly recommend this candy  i served it at a beachthemed party and everyone loved it"
]

# Split data into tokens
documents = [doc.split() for doc in data]

# Compute TF for each document
tf_data = [compute_tf(doc) for doc in documents]

# Create DataFrame for TF
tf_df = pd.DataFrame(tf_data).fillna(0)
print("TF Scores:")
print(tf_df)

# Compute IDF
idf = compute_idf(documents)
idf_df = pd.DataFrame([idf]).fillna(0)
print("\nIDF Scores:")
print(idf_df)

# Compute TF-IDF for each document
tfidf_data = [compute_tfidf(doc, idf) for doc in documents]

# Create DataFrame for TF-IDF
tfidf_df = pd.DataFrame(tfidf_data).fillna(0)
print("\nTF-IDF Scores:")
print(tfidf_df)

TF Scores:
       this     taffy        is        so      good        it      very  \
0  0.041667  0.041667  0.083333  0.041667  0.041667  0.083333  0.083333   
1  0.040000  0.000000  0.040000  0.000000  0.080000  0.000000  0.040000   
2  0.040816  0.020408  0.000000  0.000000  0.000000  0.040816  0.020408   

       soft       and     chewy  ...        in  expensive   version  \
0  0.041667  0.041667  0.041667  ...  0.000000   0.000000  0.000000   
1  0.000000  0.000000  0.000000  ...  0.000000   0.000000  0.000000   
2  0.020408  0.061224  0.020408  ...  0.020408   0.020408  0.020408   

   fralingers    highly    served  beachthemed     party  everyone     loved  
0    0.000000  0.000000  0.000000     0.000000  0.000000  0.000000  0.000000  
1    0.000000  0.000000  0.000000     0.000000  0.000000  0.000000  0.000000  
2    0.020408  0.020408  0.020408     0.020408  0.020408  0.020408  0.020408  

[3 rows x 67 columns]

IDF Scores:
         at       for     every  required      eats

# Model Selection (Lexicon Approach)

In [21]:
from textblob import TextBlob
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from tabulate import tabulate
from nltk.sentiment import SentimentIntensityAnalyzer
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import TfidfVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import LinearSVC
from sklearn.metrics import accuracy_score, classification_report

# 1. Lexicon-based approach using NLTK's VADER:

In [22]:
# Download the VADER lexicon
nltk.download('vader_lexicon')

# Assign sentiment labels based on the 'Score' column
def assign_sentiment(score):
    if score >= 4:
        return 'Positive'
    elif score <= 2:
        return 'Negative'
    else:
        return 'Neutral'

df['Sentiment'] = df['Score'].apply(assign_sentiment)

# Initialize the sentiment analyzer
sid = SentimentIntensityAnalyzer()

# Calculate sentiment scores for each review
df['Lexicon_Sentiment'] = df['clean_number'].apply(lambda x: sid.polarity_scores(x)['compound'])

# Map sentiment scores to labels
df['Lexicon_Sentiment_Label'] = df['Lexicon_Sentiment'].apply(lambda x: 'Positive' if x > 0 else ('Negative' if x < 0 else 'Neutral'))

# Evaluate the lexicon-based approach
lexicon_accuracy = accuracy_score(df['Sentiment'], df['Lexicon_Sentiment_Label'])
print("Lexicon-based Approach Accuracy:", lexicon_accuracy)

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     C:\Users\Marsya\AppData\Roaming\nltk_data...


Lexicon-based Approach Accuracy: 0.7990796089041505


In [23]:
data = [
    ("good flavor these came securely packed they were fresh and delicious i love these twizzlers", 'positive'),
    ("this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste", 'negative'),
    ("the flavors are good however i do not see any differce between this and oaker oats brand they are both mushy", 'neutral'),
    ("the strawberry twizzlers are my guilty pleasure yummy six pounds will be around for a while with my son and i", 'positive'),
    ("this oatmeal is not good its mushy soft i don t like it quaker oats is the way to go", 'negative'),
    ("halloween is over but i sent a bag to my daughters class for her share the chocolate was fresh and enjoyed by many", 'neutral'),
    ("my daughter loves twizzlers and this shipment of six pounds really hit the spot it s exactly what you would expect six packages of strawberry twizzlers", 'positive'),
    ("arrived in days and were so stale i could not eat any of the bags", 'negative'),
    ("this bag of candy online is pretty expensive it should be cheaper in order to compete with grocery stores other than that its a good combination of my favorite candy", 'neutral'),
    ("i love eating them and they are good for watching tv and looking at movies it is not too sweet i like to transfer them to a zip lock baggie so they stay fresh so i can take my time eating them", 'positive'),
    ("no tea flavor at all just whole brunch of artifial flavors it is not returnable i wasted bucks", 'negative'),
    ("this stuff really works for preventing cramping during the middle to latter stages of your rides pop into each water bottle and you re set flavor is fine and goes down easy", 'neutral'),
    ("i fed this to my golden retriever and he hated it he wouldn t eat it and when he did it gave him terrible diarrhea we will not be buying this again it s also super expensive", 'negative'),
    ("i am very satisfied with my twizzler purchase i shared these with others and we have all enjoyed them i will definitely be ordering more", 'positive'),
    ("it is okay i would not go out of my way to buy it again", 'neutral'),
    ("candy was delivered very fast and was purchased at a reasonable price i was home bound and unable to get to a store so this was perfect for me", 'positive'),
    ("not what i was expecting in terms of the company s reputation for excellent home delivery products", 'negative'),
    ("this is the same food we get at pet store but it s delivered to my door and for the same price or slightly less ", 'neutral'),
    ("i can remember buying this candy as a kid and the quality hasn t dropped in all these years still a superb product you won t be disappointed with", 'positive'),
    ("i was disappointed in the flavor and texture of this mix i usually like most of the low carb things i have tried but was diappointed in this specific one ", 'negative'),
    ("i ve been eating ramen noodles since i was a little kid and i ve never found a better flavor than hot spicy chicken it isn t hot at all to a chilihead like me but it sure is good", 'neutral'),
    ("this was sooooo deliscious but too bad i ate em too fast and gained pds my fault", 'positive'),
    ("besides being smaller than runts they look the same and have the same consistency unfortunately they taste nothing like banana runts nor do they even taste good yucky stuff trying to return with vendor", 'negative'),
    ("the item over all was fine but the banana heads are not just like banana runts they are alot smaller also when i received the item the packaging was not great the banana heads had come open during shipping and were all over the packing envelope", 'neutral'),
]


In [24]:
table_data = [["Text", "Actual Label", "TextBlob Polarity", "TextBlob Sentiment", "VADER Compound", "VADER Sentiment"]]

In [25]:
for text, actual_label in data:
    # TextBlob
    blob = TextBlob(text)
    tb_polarity = blob.sentiment.polarity
    
    # Determine label based on polarity score from TextBlob
    if tb_polarity > 0:
        tb_label = 'positive'
    elif tb_polarity < 0:
        tb_label = 'negative'
    else:
        tb_label = 'neutral'

    # VADER
    analyzer = SentimentIntensityAnalyzer()
    vs = analyzer.polarity_scores(text)
    vader_compound = vs['compound']

    # Determine label based on compound score from VADER
    if vader_compound > 0.05:
        vader_label = 'positive'
    elif vader_compound < -0.05:
        vader_label = 'negative'
    else:
        vader_label = 'neutral'

    table_data.append([text, actual_label, tb_polarity, tb_label, vader_compound, vader_label])

In [26]:
print(tabulate(table_data, headers="firstrow", tablefmt="plain"))

Text                                                                                                                                                                                                                                                  Actual Label      TextBlob Polarity  TextBlob Sentiment      VADER Compound  VADER Sentiment
good flavor these came securely packed they were fresh and delicious i love these twizzlers                                                                                                                                                           positive                 0.58        positive                        0.9382  positive
this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste                                                                 negative                 0.05        positive                        0.2363  positi

In [27]:
from textblob import TextBlob
from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer
from sklearn.metrics import classification_report
from tabulate import tabulate

data = [
    ("good flavor these came securely packed they were fresh and delicious i love these twizzlers", 'positive'),
    ("this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste", 'negative'),
    ("the flavors are good however i do not see any differce between this and oaker oats brand they are both mushy", 'neutral'),
    ("the strawberry twizzlers are my guilty pleasure yummy six pounds will be around for a while with my son and i", 'positive'),
    ("this oatmeal is not good its mushy soft i don t like it quaker oats is the way to go", 'negative'),
    ("halloween is over but i sent a bag to my daughters class for her share the chocolate was fresh and enjoyed by many", 'neutral'),
    ("my daughter loves twizzlers and this shipment of six pounds really hit the spot it s exactly what you would expect six packages of strawberry twizzlers", 'positive'),
    ("arrived in days and were so stale i could not eat any of the bags", 'negative'),
    ("this bag of candy online is pretty expensive it should be cheaper in order to compete with grocery stores other than that its a good combination of my favorite candy", 'neutral'),
    ("i love eating them and they are good for watching tv and looking at movies it is not too sweet i like to transfer them to a zip lock baggie so they stay fresh so i can take my time eating them", 'positive'),
    ("no tea flavor at all just whole brunch of artifial flavors it is not returnable i wasted bucks", 'negative'),
    ("this stuff really works for preventing cramping during the middle to latter stages of your rides pop into each water bottle and you re set flavor is fine and goes down easy", 'neutral'),
    ("i fed this to my golden retriever and he hated it he wouldn t eat it and when he did it gave him terrible diarrhea we will not be buying this again it s also super expensive", 'negative'),
    ("i am very satisfied with my twizzler purchase i shared these with others and we have all enjoyed them i will definitely be ordering more", 'positive'),
    ("it is okay i would not go out of my way to buy it again", 'neutral'),
    ("candy was delivered very fast and was purchased at a reasonable price i was home bound and unable to get to a store so this was perfect for me", 'positive'),
    ("not what i was expecting in terms of the company s reputation for excellent home delivery products", 'negative'),
    ("this is the same food we get at pet store but it s delivered to my door and for the same price or slightly less ", 'neutral'),
    ("i can remember buying this candy as a kid and the quality hasn t dropped in all these years still a superb product you won t be disappointed with", 'positive'),
    ("i was disappointed in the flavor and texture of this mix i usually like most of the low carb things i have tried but was diappointed in this specific one ", 'negative'),
    ("i ve been eating ramen noodles since i was a little kid and i ve never found a better flavor than hot spicy chicken it isn t hot at all to a chilihead like me but it sure is good", 'neutral'),
    ("this was sooooo deliscious but too bad i ate em too fast and gained pds my fault", 'positive'),
    ("besides being smaller than runts they look the same and have the same consistency unfortunately they taste nothing like banana runts nor do they even taste good yucky stuff trying to return with vendor", 'negative'),
    ("the item over all was fine but the banana heads are not just like banana runts they are alot smaller also when i received the item the packaging was not great the banana heads had come open during shipping and were all over the packing envelope", 'neutral'),
]


# Initialize an empty list to store the data in tabular format
table_data = [["Text", "Actual Label", "TextBlob Polarity", "TextBlob Sentiment", "VADER Compound", "VADER Sentiment"]]

# Lexicon-based approach using TextBlob and VADER
for text, actual_label in data:
    # TextBlob
    blob = TextBlob(text)
    tb_polarity = blob.sentiment.polarity
    
    # Determine label based on polarity score from TextBlob
    if tb_polarity > 0:
        tb_label = 'positive'
    elif tb_polarity < 0:
        tb_label = 'negative'
    else:
        tb_label = 'neutral'

    # VADER
    analyzer = SentimentIntensityAnalyzer()
    vs = analyzer.polarity_scores(text)
    vader_compound = vs['compound']
   
    # Determine label based on compound score from VADER
    if vader_compound > 0.05:
        vader_label = 'positive'
    elif vader_compound < -0.05:
        vader_label = 'negative'
    else:
        vader_label = 'neutral'

    table_data.append([text, actual_label, tb_polarity, tb_label, vader_compound, vader_label])

print(tabulate(table_data, headers="firstrow", tablefmt="plain"))

# Calculate classification report for TextBlob
tb_actual_labels = [label for _, label, _, _, _, _ in table_data[1:]]
tb_predicted_labels = [tb_label for _, _, _, tb_label, _, _ in table_data[1:]]
tb_classification_report = classification_report(tb_actual_labels, tb_predicted_labels, target_names=['negative', 'neutral', 'positive'])

# Calculate classification report for VADER
vader_actual_labels = [label for _, label, _, _, _, _ in table_data[1:]]
vader_predicted_labels = [vader_label for _, _, _, _, _, vader_label in table_data[1:]]
vader_classification_report = classification_report(vader_actual_labels, vader_predicted_labels, target_names=['negative', 'neutral', 'positive'])

# Print classification report for TextBlob
print("\nClassification Report for TextBlob:")
print(tb_classification_report)

# Print classification report for VADER
print("\nClassification Report for VADER:")
print(vader_classification_report)

Text                                                                                                                                                                                                                                                  Actual Label      TextBlob Polarity  TextBlob Sentiment      VADER Compound  VADER Sentiment
good flavor these came securely packed they were fresh and delicious i love these twizzlers                                                                                                                                                           positive                 0.58        positive                        0.9382  positive
this is the same stuff you can buy at the big box stores there is nothing healthy about it it is just carbs and sugars save your money and get something that at least has some taste                                                                 negative                 0.05        positive                        0.2363  positi

# Model Selection (Machine Learning Approach)

In [28]:
from sklearn.model_selection import train_test_split
from sklearn.feature_extraction.text import CountVectorizer
from sklearn.naive_bayes import MultinomialNB
from sklearn.svm import SVC
from sklearn.metrics import classification_report

In [29]:
# Split the data into training and testing sets
X_train, X_test, y_train, y_test = train_test_split(df['clean_number'], df['Sentiment'], test_size=0.2, random_state=42)

# Create a TF-IDF vectorizer
tfidf = TfidfVectorizer()

# Fit and transform the training data
X_train_tfidf = tfidf.fit_transform(X_train)

# Transform the testing data
X_test_tfidf = tfidf.transform(X_test)

# Train and evaluate Naive Bayes classifier
nb_classifier = MultinomialNB()
nb_classifier.fit(X_train_tfidf, y_train)
nb_predictions = nb_classifier.predict(X_test_tfidf)
nb_accuracy = accuracy_score(y_test, nb_predictions)
print("Naive Bayes Accuracy:", nb_accuracy)
print(classification_report(y_test, nb_predictions))

# Train and evaluate SVM classifier
svm_classifier = LinearSVC()
svm_classifier.fit(X_train_tfidf, y_train)
svm_predictions = svm_classifier.predict(X_test_tfidf)
svm_accuracy = accuracy_score(y_test, svm_predictions)
print("SVM Accuracy:", svm_accuracy)
print(classification_report(y_test, svm_predictions))

Naive Bayes Accuracy: 0.7917337344204907
              precision    recall  f1-score   support

    Negative       0.95      0.06      0.12     16181
     Neutral       0.41      0.00      0.00      8485
    Positive       0.79      1.00      0.88     89025

    accuracy                           0.79    113691
   macro avg       0.72      0.35      0.33    113691
weighted avg       0.78      0.79      0.71    113691

SVM Accuracy: 0.8923661503549094
              precision    recall  f1-score   support

    Negative       0.78      0.76      0.77     16181
     Neutral       0.68      0.31      0.42      8485
    Positive       0.92      0.97      0.95     89025

    accuracy                           0.89    113691
   macro avg       0.79      0.68      0.71    113691
weighted avg       0.88      0.89      0.88    113691



## Discussion 

In this section, we will be discussing about the strengths and weaknesses of the selected models for sentiment classification using the dataset about reviews of fine foods from Amazon.

- Lexicon-Based Approach Accuracy: 0.7990796089041505
- Naive Bayes Accuracy: 0.7917337344204907
- SVM Accuracy:  0.8923661503549094

1. Lexicon-Based Approach (VADER)

  STRENGTHS
  - This approach boasts a high accuracy indicating LBA is effective in classifying sentiment based on predefined word lists.
  - It is also easy to understand and implement by simply identifying which words contribute to the respective sentiments and performs well for basic sentiment analysis tasks.
  - VADER provides a compound score indicating the sentiment intensity, providing a more refined understanding of the expressed sentiments.

  WEAKNESSES
  - Vader can handle some informal language but it might struggle with complex sentence structures, sarcasm or heavily context-dependent sentiment.
  - Vader relies solely on word presence where it might miss the impact of word order, negation or irony on sentiment. 
  - The effectiveness of Vader depends on the text domain where it might not be as accurate for formal writing ot technical writing.

2. ML-Based Approach (Naive Bayes)

  STRENGTHS
  - It is one of the simple algoritm to understand and implement. Training this type of model is computationally efficient which is ideal for tasks where fast training on large datasets is crucial.
  - It scales will with increasing data sized as the model can efficiently retrain wihtout significant performance degradation.
  - It can handle both continuous and discrete data types effectively which is useful for sentiment analysis.

  WEAKNESSES
  - It records the lowest accuracy compared to LBA and SVM classifier in this model.
  - The performance of Naive Bayes can highly depend on feature selection and preprocessing. Choosing the right features and appropriate data preparation can impact its accuracy.
  - It might struggle to accurately identify the less frequent class depending on the data.

2. ML-Based Approach (SVM classifier)

  STRENGTHS
  - The SVM classifier demonstrates the highest accuracy among the models, proving its ability to handle complex relationship between features. This allows them to learn intricate patterns in the data which leads to more precise classifications.
  - SVMs can utilize kernel functions to map data into higher dimension and handle non-linear relationships between features which is valuable for sentiment analysis. 
  - SVMs can learn decision boundaties are more robust to noise and outliers in the data by maximizing margin which is increasing the confidence level of the classification.

WEAKNESSES
- SVMs can overfit the training data that can lead to poor performance on unseen data. 
- It is more difficult to understand and interpret compared to other models.
- It consumes more training time as the it deals with large datasets or complex kernel functions which is not ideal for real-time applications.